### ISSUE TITLE
Using step button in debugger resets editor mouse position to mid window.

### ISSUE THREAD

**Issue by invisiblesun** [godotengine/godot#45945]
**Godot version:**

**OS/device including version:**

MacOs

**Issue description:**

When using the debugger on a game instance that captures the mouse, step button use centers the editor mouse in the editor window.

For example a bit of code that toggles it:

func cursor_toggle():
	if Input.is_action_pressed("ui_cancel"):
		if Input.get_mouse_mode() == Input.MOUSE_MODE_VISIBLE:
			Input.set_mouse_mode(Input.MOUSE_MODE_CAPTURED)
		else:
			Input.set_mouse_mode(Input.MOUSE_MODE_VISIBLE)

func _unhandled_input(_event): cursor_toggle()

**Steps to reproduce:**
Set a break point.
Capture Mouse in a game instance in _ready() or like above.
Trigger break point.
Attempt to step through script using debugger buttons.

**Minimal reproduction project:**

**Comment by invisiblesun** [comment]
If a minimal project is necessary I will produce one but i think you can drop in the few lines of code without much trouble

**Comment by Calinou** [comment]
This likely happens because when the mouse is captured, its position is centered every frame. When you set the mouse mode to `Input.MOUSE_MODE_VISIBLE` again, it will start from the center.

This can be worked around by storing the position of the mouse *before* it was captured and restoring it using `Input.warp_mouse()` after setting `Input.MOUSE_MODE_VISIBLE`. See https://github.com/godotengine/godot/pull/39625 for an example of this.

**Comment by invisiblesun** [comment]
The -editor- is effected, not the game window, where I don't care.

When I have the mouse captured, I don't care where it is in the game window.

Maybe I missed something here, but the game mouse settings should not be effecting the editor window, right?

**Comment by invisiblesun** [comment]
Lastly, I know I can use function keys, where it isn't an issue but the effect is just weird.

**Comment by Calinou** [comment]
> Maybe I missed something here, but the game mouse settings should not be effecting the editor window during debugging, right?

Actually, it can. The mouse cursor position is a "global" variable of sorts, not a per-window variable. If you use captured mouse mode, it will center your mouse position on the screen regardless of the window or process that captured the mouse.

**Comment by invisiblesun** [comment]
Already using the edit function. I understand.

Regarding the global-ness OK.

**Comment by Zylann** [comment]
Still valid in Godot 3.4.
[DebuggerStepLocksMouse.zip](https://github.com/godotengine/godot/files/7575034/DebuggerStepLocksMouse.zip)
Place breakpoint in main.gd, launch game, press space, step with either the button or F10.

I'm not sure why the mouse is supposed to be locked "due to one frame being run", because that should not be the case. We are stepping through lines of a script within a frame of the game without even giving back focus to the game window, so I don't see any reason for the mouse to be warped.

**Comment by rk042** [comment]
> This likely happens because when the mouse is captured, its position is centered every frame. When you set the mouse mode to `Input.MOUSE_MODE_VISIBLE` again, it will start from the center.
> 
> This can be worked around by storing the position of the mouse _before_ it was captured and restoring it using `Input.warp_mouse()` after setting `Input.MOUSE_MODE_VISIBLE`. See [#39625](https://github.com/godotengine/godot/pull/39625) for an example of this.

That’s correct — under normal use this doesn’t occur. However, when code is used to manage input.mouse_mode, we do encounter the issue.
